pkg/kf/cli/interactive_command.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"errors"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+// Runner is invoked if the InteractiveNode comes up in the decision digraph.
+// What it returns guides the interactive experience.
+type Runner func(context.Context, *cobra.Command, []string) (*InteractiveNode, error)
+
+// InteractiveNode is a node of the decision digraph.
+type InteractiveNode struct {
+	// Setup is ran while traversing the digraph. Its Runner is only invoked
+	// if it comes up in the decision digraph. The returned InteractiveNodes
+	// are ALL the possible paths the interactive tutorial can go.
+	Setup func(*pflag.FlagSet) ([]*InteractiveNode, Runner)
+
+	// ctx gets set by InteractiveWithContext(). If it's not set, the normal
+	// context is used during execution.
+	ctx context.Context
```

Comment:
Passing contexts around like this rather than using them to span function calls is a bit of an anti-pattern

Reply:
Yup, good point.

Instead I'll return the context to allow each Runner to modify it.